workflow/artifacts/git/git.go

```diff
@@ -28,130 +27,136 @@ type GitArtifactDriver struct {
+case "$1" in
+Username*) echo "${GIT_USERNAME}" ;;
+Password*) echo "${GIT_PASSWORD}" ;;
+esac
+`), 0755)
+			if err != nil {
+				return nil, nil, nil, err
+			}
+		}
+		return func() {},
+			&http.BasicAuth{Username: g.Username, Password: g.Password},
+			[]string{
+				"GIT_ASKPASS=" + filename,
+				"GIT_USERNAME=" + g.Username,
+				"GIT_PASSWORD=" + g.Password,
+			},
+			nil
 	}
-	return gitClone(path, inputArtifact, nil, "")
+	return func() {}, nil, nil, nil
 }
 
 // Save is unsupported for git output artifacts
 func (g *GitArtifactDriver) Save(string, *wfv1.Artifact) error {
-	return errors.Errorf(errors.CodeBadRequest, "Git output artifacts unsupported")
+	return errors.New("git output artifacts unsupported")
 }
 
-func writePrivateKey(key string, insecureIgnoreHostKey bool) error {
-	usr, err := user.Current()
-	if err != nil {
-		return errors.InternalWrapError(err)
-	}
-	sshDir := fmt.Sprintf("%s/.ssh", usr.HomeDir)
-	err = os.MkdirAll(sshDir, 0700)
-	if err != nil {
-		return errors.InternalWrapError(err)
-	}
-
-	if insecureIgnoreHostKey {
-		sshConfig := `Host *
-	StrictHostKeyChecking no
-	UserKnownHostsFile /dev/null`
-		err = ioutil.WriteFile(fmt.Sprintf("%s/config", sshDir), []byte(sshConfig), 0644)
-		if err != nil {
-			return errors.InternalWrapError(err)
-		}
-	}
-	err = ioutil.WriteFile(fmt.Sprintf("%s/id_rsa", sshDir), []byte(key), 0600)
+func (g *GitArtifactDriver) Load(inputArtifact *wfv1.Artifact, path string) error {
+	closer, auth, env, err := g.auth()
 	if err != nil {
-		return errors.InternalWrapError(err)
+		return err
 	}
-
-	return nil
-}
-
-func gitClone(path string, inputArtifact *wfv1.Artifact, auth transport.AuthMethod, privateKey string) error {
-	cloneOptions := git.CloneOptions{
+	defer closer()
+	repo, err := git.PlainClone(path, false, &git.CloneOptions{
 		URL:               inputArtifact.Git.Repo,
 		RecurseSubmodules: git.DefaultSubmoduleRecursionDepth,
 		Auth:              auth,
-	}
-	if inputArtifact.Git.Depth != nil {
-		cloneOptions.Depth = int(*inputArtifact.Git.Depth)
-	}
-
-	repo, err := git.PlainClone(path, false, &cloneOptions)
+		Depth:             inputArtifact.Git.GetDepth(),
+	})
 	if err != nil {
-		return errors.InternalWrapError(err)
+		return err
 	}
-
 	if inputArtifact.Git.Fetch != nil {
 		refSpecs := make([]config.RefSpec, len(inputArtifact.Git.Fetch))
 		for i, spec := range inputArtifact.Git.Fetch {
 			refSpecs[i] = config.RefSpec(spec)
 		}
-
 		fetchOptions := git.FetchOptions{
+			Auth:     auth,
 			RefSpecs: refSpecs,
+			Depth:    inputArtifact.Git.GetDepth(),
 		}
-		if inputArtifact.Git.Depth != nil {
-			fetchOptions.Depth = int(*inputArtifact.Git.Depth)
-		}
-
 		err = fetchOptions.Validate()
 		if err != nil {
-			return errors.InternalWrapError(err)
+			return err
 		}
-
 		err = repo.Fetch(&fetchOptions)
-		if err != nil {
-			return errors.InternalWrapError(err)
+		if err != nil && err.Error() != "already up-to-date" {
```

Comment:
Extract to function or constant? `IsAlreadyUpToDateErr(err)`

Reply:
good idea - there was no easy way to check this

Reply:
done